exporters/prometheus/exporter.go

```diff
@@ -344,3 +356,116 @@ func sanitizeName(n string) string {
 
 	return b.String()
 }
+
+func (c *collector) validateMetrics(m metricdata.Metrics) (bool, metricdata.Metrics, error) {
+	// get the name of metric name
+	name := c.getName(m)
+	switch m.Data.(type) {
+	case metricdata.Sum[int64]:
+		sum := m.Data.(metricdata.Sum[int64])
+		if sum.IsMonotonic {
+			name += counterSuffix
+		}
+	case metricdata.Sum[float64]:
+		sum := m.Data.(metricdata.Sum[float64])
+		if sum.IsMonotonic {
+			name += counterSuffix
+		}
+	}
+
+	metricFamily := &dto.MetricFamily{
+		Name: proto.String(name),
+		Help: proto.String(m.Description),
+	}
+	switch m.Data.(type) {
+	case metricdata.Sum[int64]:
+		sum := m.Data.(metricdata.Sum[int64])
+		metricFamily.Type = dto.MetricType_COUNTER.Enum()
+		if !sum.IsMonotonic {
+			metricFamily.Type = dto.MetricType_GAUGE.Enum()
+		}
+	case metricdata.Sum[float64]:
+		sum := m.Data.(metricdata.Sum[float64])
+		metricFamily.Type = dto.MetricType_COUNTER.Enum()
+		if !sum.IsMonotonic {
+			metricFamily.Type = dto.MetricType_GAUGE.Enum()
+		}
+	case metricdata.Gauge[int64]:
+		metricFamily.Type = dto.MetricType_GAUGE.Enum()
+	case metricdata.Gauge[float64]:
+		metricFamily.Type = dto.MetricType_GAUGE.Enum()
+	case metricdata.Histogram:
+		metricFamily.Type = dto.MetricType_HISTOGRAM.Enum()
+	}
+
+	collision := validateCollision(metricFamily, c.metricFamiliesByName)
+	switch collision {
+	case noCollision:
+		c.metricFamiliesByName[name] = metricFamily
+		return false, m, nil
+	case typeCollision:
+		return true, m, fmt.Errorf("type collision")
+	case helpCollision:
+		emf := c.metricFamiliesByName[name]
+		m.Description = emf.GetHelp()
+		return false, m, fmt.Errorf("help collision")
+	}
+
+	return false, m, nil
+}
+
+type collisionType int
+
+const (
+	noCollision   collisionType = 0
+	typeCollision collisionType = 1
+	helpCollision collisionType = 2
+)
+
+func validateCollision(mf *dto.MetricFamily, mfs map[string]*dto.MetricFamily) collisionType {
```

Comment:
We don't need to duplicate the existing prom metric validation logic (checking that suffixed metrics don't collide with non-suffixed metrics).  We just need to validate that the descriptions don't collide, which doesn't involve checking suffixes.

Reply:
as #2890 said
> Each time a new metric (with Description) is added, look up the name in the map:
> - If it isn't found, add it to the map
> - If it is found, and the desc is the same, do nothing
> - If it is found, and has conflicting type information, drop the metric + log a warning
> - If it is found, and has conflicting help or unit, use the stored help + unit instead of the provided one + log a warning.

If we do not check that suffixed metrics don't collide with non-suffixed metrics, how do we handle the case like following
```
histogram, err := meterA.SyncInt64().Histogram("foo",
					instrument.WithUnit(unit.Bytes),
					instrument.WithDescription("meter foo"))
gauge, err := meterA.SyncInt64().UpDownCounter("foo_count",
					instrument.WithUnit(unit.Bytes),
					instrument.WithDescription("meter foo"))
```
the two instrument both produces metric `foo_count`, and in the prometheus sdk, this case will failed the metric scrape request.

@dashpole